src/transformers/models/kosmos2/processing_kosmos2.py

```diff
@@ -0,0 +1,608 @@
+# coding=utf-8
+# Copyright 2023 Microsoft Research and The HuggingFace Inc. team. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Processor class for KOSMOS-2."""
+
+import copy
+import math
+import re
+from typing import List, Optional, Tuple, Union
+
+import numpy as np
+
+from ...image_processing_utils import BatchFeature
+from ...image_utils import ImageInput, is_batched
+from ...processing_utils import ProcessorMixin
+from ...tokenization_utils_base import PaddingStrategy, TextInput, TruncationStrategy
+from ...utils import TensorType, is_torch_available
+
+
+BboxInput = Union[
+    List[Tuple[int, int]],
+    List[Tuple[float, float, float, float]],
+    List[List[Tuple[int, int]]],
+    List[List[Tuple[float, float, float]]],
+]
+
+
+class Kosmos2Processor(ProcessorMixin):
+    r"""
+    Constructs an KOSMOS-2 processor which wraps a KOSMOS-2 image processor and a KOSMOS-2 tokenizer into a single
+    processor.
+
+    [`Kosmos2Processor`] offers all the functionalities of [`Kosmos2ImageProcessor`] and some functionalities of
```

Comment:
`Kosmos2Processor` applies some changes to the input string (adding some extra tokens related to image information, etc.), and it needs to return a `image_features_mask`.

It also allows to add `EOS` token depending on if we are doing generation or not. For this to work correctly between the slow/fast tokenizer, we need to add `BOS` token or not at string level.

In order to make the above 2 things work correctly together with padding, it calls the `tokenizer` without padding, compute `image_features_mask`, then adding pad tokens if necessary.

However, this can't provide all the (huge number)  functionalities our base tokenizer class provides.

Reply:
See https://huggingface.slack.com/archives/C01N44FJDHT/p1693836780294819?thread_ts=1693828912.634649&cid=C01N44FJDHT